types/src/transaction/authenticator.rs

```diff
@@ -76,84 +84,184 @@ impl TransactionAuthenticator {
         }
     }
 
+    /// Return the raw bytes of `self.public_key`
     pub fn public_key_bytes(&self) -> Vec<u8> {
         match self {
             Self::Ed25519 { public_key, .. } => public_key.to_bytes().to_vec(),
             Self::MultiEd25519 { public_key, .. } => public_key.to_bytes().to_vec(),
         }
     }
 
+    /// Return the raw bytes of `self.signature`
     pub fn signature_bytes(&self) -> Vec<u8> {
         match self {
             Self::Ed25519 { signature, .. } => signature.to_bytes().to_vec(),
             Self::MultiEd25519 { signature, .. } => signature.to_bytes().to_vec(),
         }
     }
 
-    /// Return bytes for (self.public_key | self.scheme_id). To create an account authentication
-    /// key, sha3 these bytes
-    // TODO: move AuthenticationKey to this file and make this private
-    pub fn compute_authentication_key_preimage(
-        mut public_key_bytes: Vec<u8>,
-        scheme_id: u8,
-    ) -> AuthenticationKeyPreimage {
-        public_key_bytes.push(scheme_id);
-        AuthenticationKeyPreimage(public_key_bytes)
+    /// Return an authentication key preimage derived from `self`'s public key and scheme id
+    pub fn authentication_key_preimage(&self) -> AuthenticationKeyPreimage {
+        AuthenticationKeyPreimage::new(self.public_key_bytes(), self.scheme())
     }
 
-    // TODO: move AuthenticationKey to this file to avoid try_from
-    fn preimage_to_authentication_key(preimage: &AuthenticationKeyPreimage) -> AuthenticationKey {
-        AuthenticationKey::try_from(HashValue::from_sha3_256(&preimage.0).to_vec()).unwrap()
+    /// Return an authentication key derived from `self`'s public key and scheme id
+    pub fn authentication_key(&self) -> AuthenticationKey {
+        AuthenticationKey::from_preimage(&self.authentication_key_preimage())
     }
+}
 
-    pub fn authentication_key_preimage(&self) -> AuthenticationKeyPreimage {
-        Self::compute_authentication_key_preimage(
-            self.public_key_bytes().to_vec(),
-            self.scheme_id(),
-        )
+/// A struct that represents an account authentication key. An account's address is the last 16
+/// bytes of authentication key used to create it
+#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Hash, Clone, Copy, CryptoHasher)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+pub struct AuthenticationKey([u8; AUTHENTICATION_KEY_LENGTH]);
```

Comment:
can we make the AUTHENTICATION_KEY_LENGTH a const in AuthenticationKey struct? It seems more natural for it to be AuthenticationKey::LENGTH.